CORE-13920 message pattern lib metrics

conal-smith opened the pull request:
I've now updated [flow dashboard](https://github.com/corda/engineering-helm/pull/50/files#diff-f4bd8bfd6097f7b6628b3835244f804d8173df626065098302e11e3ff6ebc576) following renaming of some "messaging" metrics to "consumer" (as they cover all consumers).

See grafana dashboard of flow worker with these updates [here](https://grafana.non-functional-tests.corda.cloud/d/Conal-FlowWorker/conal-flow-worker?orgId=1&from=1687544802980&to=1687545779676).

See the message pattern dashboard [here](https://grafana.non-functional-tests.corda.cloud/d/Conal-Messaging/conal-messaging-patterns?orgId=1&from=1687544779198&to=1687545767527)

See engineering-helm changes [here](https://github.com/corda/engineering-helm/pull/50).

conalsmith-r3 commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/consumer/CordaKafkaConsumerImpl.kt:
```diff
@@ -114,7 +121,22 @@ class CordaKafkaConsumerImpl<K : Any, V : Any>(
 
-        return recordsToReturn.sortedBy { it.timestamp }
+        return recordsToReturn.sortedBy { it.timestamp }.also {
+            CordaMetrics.Metric.Messaging.ConsumerPollTime.builder()
```
This is consumer poll time for all consumers, previously we had it only for S&E consumers. I haven't removed `MessagingPollTime`, but could do so if you're happy with that @JamesHR3?

sprintengine replied:
This replaces `MessagingPollTime`

sprintengine commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/consumer/CordaKafkaConsumerImpl.kt:
```diff
@@ -114,7 +121,22 @@ class CordaKafkaConsumerImpl<K : Any, V : Any>(
+
+    private fun recordPolledRecordsPerPartition(partition: Int, records: List<ConsumerRecord<*, *>>) {
+        recordsConsumedMetricCache.computeIfAbsent(partition) {
+            CordaMetrics.Metric.Messaging.ConsumerRecordsConsumedCount.builder()
```
Removed `messaging.batch.size` in place of this.

sprintengine commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/producer/CordaKafkaProducerImpl.kt:
```diff
@@ -140,12 +140,28 @@ class CordaKafkaProducerImpl(
+                    .record(count.toDouble())
+            }
+
+    }
```
Tested this locally with `SendReceiveAllMessagingFlow` and saw that it works, the current perf test suites don't create msgs large enough to send chunked.

conalsmith-r3 commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/producer/CordaKafkaProducerImpl.kt:
```diff
@@ -140,12 +140,28 @@ class CordaKafkaProducerImpl(
+            .forEach { (topic, count) ->
+                CordaMetrics.Metric.Messaging.ProducerChunksGenerated.builder()
+                    .withTag(CordaMetrics.Tag.MessagePatternClientId, config.clientId)
+                    .withTag(CordaMetrics.Tag.Topic, topic)
```
arguably we may not need topic here, if we have the clientId, and we could build the metric at class level rather than rebuilding each time chunking is used.

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/StateAndEventConsumerImpl.kt:
```diff
@@ -59,13 +61,31 @@ internal class StateAndEventConsumerImpl<K : Any, S : Any, E : Any>(
+                    .withTag(CordaMetrics.Tag.Partition, "$partition")
+                    .build()
+            }
+                .record((currentStates[partition]?.keys?.size ?: 0).toDouble())
```
`currentStates` is a map of partitions to a map of states, where the key is the state key and the value is a pair of the timestamp and the state.

For the given partition we record the number of state keys.

Though this seems to have some unexpected data:
- states increasing by exactly 6 each time for each partition
- all partitions have exactly the same number of states

![image](https://github.com/corda/corda-runtime-os/assets/68279385/198c352b-0c09-4056-b18e-06a2e452a101)

sprintengine commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -614,6 +594,51 @@ object CordaMetrics {
+            class PartitionedConsumerInMemoryStore(computation: Supplier<Number>) : ComputedValue<Nothing>(
+                "consumer.partitioned.inmemory.store",
+                computation
+            )
```
Micrometer metrics don't like when you use one metric name built in two consumers where one sets a tag ("partition") and the other doesn't, so I have two separate ones to avoid confusion

conalsmith-r3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/StateAndEventConsumerImpl.kt:
```diff
@@ -60,18 +61,20 @@ internal class StateAndEventConsumerImpl<K : Any, S : Any, E : Any>(
-    private val eventPollTimer = CordaMetrics.Metric.MessagePollTime.builder()
-        .withTag(CordaMetrics.Tag.MessagePatternType, MetricsConstants.STATE_AND_EVENT_PATTERN_TYPE)
-        .withTag(CordaMetrics.Tag.MessagePatternClientId, config.clientId)
-        .withTag(CordaMetrics.Tag.OperationName, MetricsConstants.EVENT_POLL_OPERATION)
```
I think both event and state consumers might have clientId "eventConsumer" so if we change the correct one to be "stateConsumer" we can discern between state polls and event polls

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/StateAndEventConsumerImpl.kt:
```diff
@@ -60,18 +61,20 @@ internal class StateAndEventConsumerImpl<K : Any, S : Any, E : Any>(
-        .withTag(CordaMetrics.Tag.MessagePatternType, MetricsConstants.STATE_AND_EVENT_PATTERN_TYPE)
-        .withTag(CordaMetrics.Tag.MessagePatternClientId, config.clientId)
-        .withTag(CordaMetrics.Tag.OperationName, MetricsConstants.STATE_POLL_OPERATION)
-        .build()
```
We can discern poll operation type "event" or "state" from the client ID - State consumer has a client ID of 
```
"${config.clientId}-stateConsumer"
```
and eventConsumer:
```
"${config.clientId}-eventConsumer"
```

jujoramos reviewed (commented):
It would be good to add tests verifying that the newly added metrics work. Otherwise, LGTM.

conalsmith-r3 commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -614,6 +594,51 @@ object CordaMetrics {
+            /**
+             * Counter for the number of records a consumer has consumed.
+             */
+            object ConsumerRecordsConsumedCount : Metric<DistributionSummary>("consumer.records.consumed", Metrics::summary)
```
Might be better to remove "Count" from this name

charlieR3 commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/consumer/CordaKafkaConsumerImpl.kt:
```diff
@@ -49,11 +51,13 @@ class CordaKafkaConsumerImpl<K : Any, V : Any>(
     private var currentAssignment = mutableSetOf<Int>()
     private val bufferedRecords = mutableMapOf<Int, List<ConsumerRecord<Any, Any>>>()
 
+    private val recordsConsumedMetricCache = ConcurrentHashMap<Int, DistributionSummary>()
```
Do you need to cache these? Micrometer caches based on name a tags already.

sprintengine replied:
You're right thanks, I've removed this hashmap

charlieR3 reviewed (dismissed):
LGTM - one minor comment

conal-smith committed:
CORE-13920 - Message pattern metrics
A) Replaced the following metrics:
- `messaging.batch.size` -> `consumer.batch.size`
- `messaging.poll.time` -> `consumer.poll.time`
B) Added the following metrics:
- `consumer.compacted.inmemory.store`
- `consumer.partitioned.inmemory.store`
- `producer.chunks.generated`
C) Added the following tags:
- `topic`
- `partition`

charlieR3 reviewed (approved):
LGTM
